docs/WritingTests.asciidoc

```diff
@@ -752,17 +768,169 @@ Sending key presses over VNC is very slow so for tests which send a lot of
+then it returns true. It is also possible to check if the current console is a
+serial terminal by calling +is_serial_terminal+.
+
+Once you have selected a serial terminal, the video feed will disappear from
+the live view, however at the bottom of the live screen there is a separate
+text feed. After the test has finished you can view the serial log(s) in the
+assets tab. You will probably have two serial logs; +serial0.txt+ and
++serial_terminal.txt+.
+
+Now that you are on a serial terminal everything will start to go a lot
+faster. So much faster in fact that race conditions become a big
+issue. Generally these can be avoided by using the higher level functions such
+as +script_run+ and +script_output+.
+
+It is rarely necessary to use the lower level functions, however it helps to
+recognise problems caused by race conditions at the lower level, so please
+read the following section regardless.
+
+So if you do need to use +type_string+ and +wait_serial+ directly then try to
+use the following pattern:
+
+1) Wait for the terminal prompt to appear.
+2) Send your command
+3) Wait for your command text to be echoed by the shell (inf applicable)
+4) Send enter
+5) Wait for your command output (if applicable)
+
+To illustrate this is a snippet from the LTP test runner which uses the lower
+level commands to achieve a little bit more control.
+
+[source,perl]
+--------------------------------------------------------------------------------
+my $fin_msg    = "### TEST $test->{name} COMPLETE >>> ";
+my $cmd_text   = qq($test->{command}; echo "$fin_msg\$?");
+my $klog_stamp = "echo 'OpenQA::run_ltp.pm: Starting $test->{name}' > /dev/$serialdev";
+
+# More variables and other shit
+
+if (is_serial_terminal) {
+        script_run($klog_stamp);
+        wait_serial(serial_term_prompt(), undef, 0, no_regex => 1);
+        type_string($cmd_text);
+        wait_serial($cmd_text, undef, 0, no_regex => 1);
+        type_string("\n");
+} else {
+        # The slow path
+}
+my $test_log = wait_serial(qr/$fin_msg\d+/, $timeout, 0, record_output => 1);
+--------------------------------------------------------------------------------
+
+The first +wait_serial+ ensures that the shell prompt has appeared. If we do
+not wait for the shell prompt then it is possible that we can send input to
+whatever command was run before. In this case that command would be 'echo'
+which is used by +script_run+ to print a 'finished' message.
+
+It is possible that echo was able to print the finish message, but was then
+suspended by the OS before it could exit. In which case the test script is able
+to race ahead and start sending input which will just be dropped by echo once
+it is resumed. Waiting for the shell prompt stops this from happening.
+
+Then the test script enters our command with +type_string+ and waits for the
+command's text to be echoed back by the shell. Shell's usually echo the
+characters sent to them so that the user can see what they have typed. However
```

Comment:
and when it is not the case and why ?

Reply:
this one is just like ongoing question not directly related to PR :)

Reply:
Ok, well I have explained a bit more anyway in the docs, but a full explanation would not be relevant because it is only likely to be turned off deliberately to simplify processing SUT output which I don't think is a good idea.

Another situation where it might be turned off is if your local terminal emulator displays your keypresses. In which case the remote TTY/PTY does not need to. Another is if you are using a very stupid terminal on an embedded device which does not echo the characters. Also if you are using a very slow serial connection, the echo might interfere with the characters you are trying to send, so it is best to turn it off. None of this is very likely when you are using a Linux capable device, but I think it is helpful knowledge for understanding how terminals work.